mlir/lib/Analysis/FlatLinearValueConstraints.cpp

```diff
@@ -893,7 +893,7 @@ FlatLinearValueConstraints::FlatLinearValueConstraints(IntegerSet set,
                             set.getNumDims(), set.getNumSymbols(),
                             /*numLocals=*/0) {
   assert(operands.empty() ||
-         set.getNumInputs() == operands.size() && "operand count mismatch");
+         (set.getNumInputs() == operands.size() && "operand count mismatch"));
```

Comment:
Yes, the point here is that the braces should not change the semantics. gcc "suggests" braces to clarify. I tried to add braces without changing semantics. What you are proposing looks more correct, but is not the same as without.

Reply:
Personally, I've always been a fan of over-bracing instead of under-bracing, especially in cases where compilers have "doubts". 

IIUC, the GGC warning isn't because it doesn't know what to do, but that it recognises that this is a common bug pattern that people don't usually get the semantics right.

Adding the braces in the right place would make it easier for humans to read and all agree on the semantics, which is a good strategy regardless of warnings.

Reply:
> What you are proposing looks more correct, but is not the same as without.

It is the same semantics (though it gets there through a different intermediate path) because string literals are always true. and `(x || y) && true` is the same as `x || (y && true)` - and the former is more what people intend, and means line wrapping/formatting will align with what the user intent, etc.

It's not that GCC doesn't know what to do - and it is a common bug pattern in general, and clang warns on this case in general - but clang has a carveout for string literals (& maybe constants more generally) because in this particular context it's not bug prone - the string literal being always true makes the bug benign.

Reply:
Adjusted braces.